src/Elm/Parser/Layout.elm

```diff
@@ -2,68 +2,137 @@ module Elm.Parser.Layout exposing
     ( layout
     , layoutStrict
     , maybeAroundBothSides
+    , maybeLayout
     , onTopIndentation
     , optimisticLayout
     , positivelyIndented
     )
 
 import Combine exposing (Parser)
 import Elm.Parser.Comments as Comments
+import Elm.Parser.Node as Node
 import Elm.Parser.State as State exposing (State)
-import Parser as Core
+import Elm.Syntax.Node exposing (Node)
+import Parser as Core exposing ((|.), (|=))
 import Set
 
 
-oneOfAnyComment : List (Combine.Parser State ())
-oneOfAnyComment =
-    [ Comments.singleLineComment
-    , Comments.multilineComment
-    ]
+nonEmptyWhiteSpaceOrComment : Core.Parser (Maybe (Node String))
+nonEmptyWhiteSpaceOrComment =
+    Core.oneOf
+        [ Core.variable
+            { inner = \c -> c == ' ' || c == '\n' || c == '\u{000D}'
+            , reserved = Set.empty
+            , start = \c -> c == ' ' || c == '\n' || c == '\u{000D}'
+            }
+            |> Core.map (\_ -> Nothing)
+        , Core.oneOf
+            [ Comments.singleLineCommentCore |> Core.getChompedString
+            , Comments.multilineCommentString
+            ]
+            |> Node.parserCoreMap Just
+        ]
+
+
+addAsCommentsToState : Core.Parser (List (Node String)) -> Parser State ()
+addAsCommentsToState parser =
+    Combine.Parser
+        (\state ->
+            parser
+                |> Core.map
+                    (\newComments ->
+                        ( State.addComments newComments state, () )
+                    )
+        )
+
+
+maybeLayout : Parser State ()
+maybeLayout =
+    coreWhiteSpaceAndComments
+        |> addAsCommentsToState
+        |> verifyLayoutIndent
```

Comment:
I'm wondering whether we could combine `verifyLayoutIndent` with the previous function to avoid a `Core.map`

(Definitely not for this PR)